File: web/src/lib/money.test.ts
import { describe, expect, it } from "vitest";

import { formatMoney, parseAmount, toDecimal } from "./money";

describe("parseAmount", () => {
  it("parses exact minor units without float errors", () => {
    expect(parseAmount("0.29", "USD")).toBe(29);
    expect(parseAmount("1,250.75", "SAR")).toBe(125075);
    expect(parseAmount("1.234", "JOD")).toBe(1234);
    expect(parseAmount("1500", "JPY")).toBe(1500);
  });

  it("accepts Arabic-Indic digits", () => {
    expect(parseAmount("١٢٫٥٠", "USD")).toBe(1250);
  });

  it("rejects invalid input and excess precision", () => {
    expect(parseAmount("1.234", "USD")).toBeNull();
    expect(parseAmount("abc", "USD")).toBeNull();
    expect(parseAmount("-5", "USD")).toBeNull();
    expect(parseAmount("-5", "USD", { allowNegative: true })).toBe(-500);
  });
});

describe("toDecimal / formatMoney", () => {
  it("round-trips", () => {
    for (const minor of [0, 7, 100, 123456789]) expect(parseAmount(toDecimal(minor, "JOD"), "JOD")).toBe(minor);
    expect(toDecimal(-5, "USD")).toBe("-0.05");
  });

  it("formats with grouping and bidi isolation", () => {
    expect(formatMoney(125075, "USD", "en")).toBe("USD ⁦1,250.75⁩");
    expect(formatMoney(-500, "USD", "en", { signed: true })).toBe("USD ⁦-5.00⁩");
  });
});

File: mobile/test/core/money_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:masroof/core/money/money.dart';

void main() {
  group('Money.tryParse', () {
    test('parses decimals exactly into minor units', () {
      expect(Money.tryParse('10', 'SAR'), 1000);
      expect(Money.tryParse('10.5', 'SAR'), 1050);
      expect(Money.tryParse('0.29', 'USD'), 29);
      expect(Money.tryParse('1,250.75', 'SAR'), 125075);
      expect(Money.tryParse('1.234', 'KWD'), 1234);
      expect(Money.tryParse('1500', 'JPY'), 1500);
    });

    test('accepts Arabic-Indic digits and Arabic decimal separator', () {
      expect(Money.tryParse('١٢٫٥٠', 'SAR'), 1250);
      expect(Money.tryParse('۳۰', 'SAR'), 3000);
    });

    test('rejects invalid input and excess precision', () {
      expect(Money.tryParse('', 'SAR'), isNull);
      expect(Money.tryParse('abc', 'SAR'), isNull);
      expect(Money.tryParse('1.234', 'SAR'), isNull);
      expect(Money.tryParse('1.5', 'JPY'), isNull);
      expect(Money.tryParse('-5', 'SAR'), isNull);
      expect(Money.tryParse('1e3', 'SAR'), isNull);
    });
  });

  group('Money.toDecimal', () {
    test('formats API decimal strings', () {
      expect(Money.toDecimal(150050, 'SAR'), '1500.50');
      expect(Money.toDecimal(7, 'SAR'), '0.07');
      expect(Money.toDecimal(-5, 'SAR'), '-0.05');
      expect(Money.toDecimal(1, 'BHD'), '0.001');
      expect(Money.toDecimal(1500, 'JPY'), '1500');
    });

    test('round-trips with tryParse', () {
      for (final minor in [0, 1, 99, 100, 123456789]) {
        expect(Money.tryParse(Money.toDecimal(minor, 'KWD'), 'KWD'), minor);
      }
    });
  });

  group('Money.format', () {
    test('uses grouping, localized symbols and bidi isolates', () {
      expect(Money.format(125075, 'SAR', locale: 'en'), 'SAR \u20661,250.75\u2069');
      expect(Money.format(-125075, 'SAR', locale: 'ar'), '\u2066-1,250.75\u2069 ر.س');
      expect(Money.format(500, 'USD', locale: 'en', signed: true), 'USD \u2066+5.00\u2069');
      expect(Money.format(1234567, 'KWD', locale: 'en', withSymbol: false), '1,234.567');
    });
  });
}
